src/types.rs

```diff
@@ -271,23 +279,79 @@ impl Type {
     pub(crate) fn validate(
         &self,
         extension_registry: &ExtensionRegistry,
+        type_vars: &[TypeParam],
     ) -> Result<(), SignatureError> {
         // There is no need to check the components against the bound,
         // that is guaranteed by construction (even for deserialization)
         match &self.0 {
-            TypeEnum::Tuple(row) | TypeEnum::Sum(SumType::General { row }) => {
-                row.iter().try_for_each(|t| t.validate(extension_registry))
-            }
+            TypeEnum::Tuple(row) | TypeEnum::Sum(SumType::General { row }) => row
+                .iter()
+                .try_for_each(|t| t.validate(extension_registry, type_vars)),
             TypeEnum::Sum(SumType::Simple { .. }) => Ok(()), // No leaves there
             TypeEnum::Prim(PrimType::Alias(_)) => Ok(()),
-            TypeEnum::Prim(PrimType::Extension(custy)) => custy.validate(extension_registry),
-            TypeEnum::Prim(PrimType::Function(ft)) => ft
-                .input
-                .iter()
-                .chain(ft.output.iter())
-                .try_for_each(|t| t.validate(extension_registry)),
+            TypeEnum::Prim(PrimType::Extension(custy)) => {
+                custy.validate(extension_registry, type_vars)
+            }
+            TypeEnum::Prim(PrimType::Function(ft)) => ft.validate(extension_registry, type_vars),
+            TypeEnum::Prim(PrimType::Variable(idx, bound)) => {
+                if type_vars.get(*idx) == Some(&TypeParam::Type(*bound)) {
+                    Ok(())
+                } else {
+                    Err(SignatureError::TypeVarDoesNotMatchDeclaration {
+                        decl: type_vars.get(*idx).cloned(),
```

Comment:
This message isn't so helpful in the case where there's a `Var(n + 1)` and `n` type args (i.e. `type_vars.get(...)` is `None`)